src/crimson/osd/pg_backend.cc

```diff
@@ -766,53 +767,68 @@ PGBackend::write_iertr::future<> PGBackend::writefull(
-mocked_load_clone_obc(const auto& coid)
-{
-  return crimson::ct_error::enoent::make();
-}
-
-static auto head2clone(const hobject_t& hoid)
-{
-  // TODO: transform hoid into coid
-  return hoid;
-}
-
 PGBackend::rollback_iertr::future<> PGBackend::rollback(
-  const SnapSet &ss,
   ObjectState& os,
   const OSDOp& osd_op,
   ceph::os::Transaction& txn,
   osd_op_params_t& osd_op_params,
-  object_stat_sum_t& delta_stats)
+  object_stat_sum_t& delta_stats,
+  crimson::osd::ObjectContextRef head,
+  crimson::osd::ObjectContextLoader& obc_loader)
 {
+  const ceph_osd_op& op = osd_op.op;
+  snapid_t snapid = (uint64_t)op.snap.snapid;
   assert(os.oi.soid.is_head());
   logger().debug("{} deleting {} and rolling back to old snap {}",
-                  __func__, os.oi.soid, osd_op.op.snap.snapid);
-  return mocked_load_clone_obc(
-    head2clone(os.oi.soid)
-  ).safe_then_interruptible([](auto clone_obc) {
-    // TODO: implement me!
-    static_cast<void>(clone_obc);
-    return remove_iertr::now();
-  }, crimson::ct_error::enoent::handle([this, &os, &txn, &delta_stats] {
-    // there's no snapshot here, or there's no object.
-    // if there's no snapshot, we delete the object; otherwise, do nothing.
-    logger().debug("rollback: deleting head on {}"
-                   " because got ENOENT|whiteout on obc lookup",
-                   os.oi.soid);
-    return remove(os, txn, delta_stats, true /*whiteout*/);
-  }), mocked_load_clone_obc_ertr::assert_all{
-    "unexpected error code in rollback"
-  });
+                  __func__, os.oi.soid ,snapid);
+  hobject_t target_coid = os.oi.soid;
+  target_coid.snap = snapid;
+  bool delete_head = false;
+  return obc_loader.with_clone_obc_only<RWState::RWREAD>(
+    head, target_coid,
+    [this, &os, &txn, &delta_stats, &obc_loader, &delete_head]
+    (auto clone_obc) {
+    logger().debug("PGBackend::rollback: loaded clone_obc: {}",
+                   clone_obc->obs.oi.soid);
+    // 1) Delete current head
+    if (os.exists) {
+        txn.remove(coll->get_cid(), ghobject_t{os.oi.soid,
+                                    ghobject_t::NO_GEN, shard});
+    }
+    // 2) Clone correct snapshot into head
+    txn.clone(coll->get_cid(), ghobject_t{clone_obc->obs.oi.soid},
+                               ghobject_t{os.oi.soid});
+    //    Copy clone obc.os.oi to os.oi
+    os.oi.clear_flag(object_info_t::FLAG_WHITEOUT);
+    os.oi.copy_user_bits(clone_obc->obs.oi);
+    // TODO: 3) Calculate clone_overlaps by following overlaps
+    //          forward from rollback snapshot
+    //          https://tracker.ceph.com/issues/58263
+    return rollback_iertr::now();
+  }).handle_error_interruptible(
+    crimson::ct_error::enoent::handle(
+      [this, &os, &snapid, &txn, &delta_stats, &delete_head] {
+      // there's no snapshot here, or there's no object.
+      // if there's no snapshot, we delete the object;
+      // otherwise, do nothing.
+      delete_head = true;
+      return rollback_ertr::now();
```

Comment:
Why not `seastar::now()` (or an unerrorated but interruptible variant of it)?

Why exactly we can't do the `remove()` logic here? Dividing into two phases seems a bit problematic (e.g. needs ensuring proper life-time for the `delete_head` flag).

Reply:
I strongly agree it seems problematic, please see my previous comment to Sam.
(Note, I did not squash the commit yet and left it as DNM for easier discussion)

Reply:
I saw only this one:

> I couldn't use `remove()` with `handle_error_interruptible` constrains.

but it doesn't _exactly_ explain what was the issue. I guess `remove()` returns an errorated-and-interruptible future and I speculate that, in the handler above, you want to return pure interruptible one. If that's the case, you might need another, inner `handle_error_interruptible` to handle (`assert_all{}`?) errors just from `remove()`.

Reply:
```
return obc_loader.with_clone_obc_only<...>(...) {
...
}).handle_error_interruptible(
  crimson::ct_error::enoent::handle(
    [this, &os, &snapid, &txn, &delta_stats, &delete_head] {
    return remove(os, txn, delta_stats, false).handle_error_interruptible(
      crimson::ct_error::assert_all{});
  }),
  crimson::osd::ObjectContextLoader::load_obc_ertr::all_same_way(
  [target_coid](auto& code) {
    ...
}));
```

`handle_error_interruptible` expects to return `errorator`:
```
../src/crimson/osd/pg_backend.cc:808:32:   required from here
../src/crimson/common/errorator.h:339:88: error: no type named ‘errorator_type’ in ‘using type = using type = class seastar::future<>’ {aka ‘class seastar::future<>’} 
```
(l.808 is `}).handle_error_interruptible(`) mentioned here.

Reply:
I ended up handling `enoent` and `assert_all` errors separately by `safe_then_interruptible`.